atlasdb-client/src/main/java/com/palantir/atlasdb/util/TaggedMetricsInvocationEventHandler.java

```diff
@@ -47,17 +52,36 @@
 
     private final TaggedMetricRegistry taggedMetricRegistry;
     private final String serviceName;
-
-    private final Function<InvocationContext, Map<String, String>> tagFunction;
+    private final Meter globalFailureMeter;
+    private final LoadingCache<Method, Timer> timerCacheByMethod;
 
     public TaggedMetricsInvocationEventHandler(
             TaggedMetricRegistry taggedMetricRegistry,
             String serviceName,
-            Function<InvocationContext, Map<String, String>> tagFunction) {
+            Function<Method, Map<String, String>> tagFunction) {
         super(InstrumentationUtils.getEnabledSupplier(serviceName));
         this.taggedMetricRegistry = Preconditions.checkNotNull(taggedMetricRegistry, "metricRegistry");
         this.serviceName = Preconditions.checkNotNull(serviceName, "serviceName");
-        this.tagFunction = tagFunction;
+        this.globalFailureMeter = initializeGlobalFailureMeter();
+        this.timerCacheByMethod = Caffeine.newBuilder()
+                .maximumSize(1000)
+                .build(new CacheLoader<Method, Timer>() {
+                    @Nonnull
+                    @Override
+                    public Timer load(@Nonnull Method key) {
+                        MetricName finalMetricName = MetricName.builder()
+                                .safeName(MetricRegistry.name(serviceName, key.getName()))
+                                .putAllSafeTags(tagFunction.apply(key))
+                                .build();
+                        return taggedMetricRegistry.timer(finalMetricName);
+                    }
+                });
```

Comment:
I don't think it makes sense to bound this because every MetricName will also be held in the TaggedMetricRegistry map. The methods are bounded by the proxy implemented interfaces.
Based on published benchmarks, caffeine has overhead beyond the concurrent hash map: https://github.com/ben-manes/caffeine/wiki/Benchmarks#read-100-1

Reply:
In general this approach doesn't work, because we have metrics functions that are not pure functions :(

Reply:
ack 👍

Reply:
Yeah, the argument for not bounding makes sense - will switch to CHM if we continue pursuing this cache approach

Reply:
actually, I'm not sure I follow, can you provide an example of a case that doesn't work to help me understand?

Reply:
So currently say for `TimestampService`, we have a tag `isCurrentSuspectedLeader` that indicates whether the node thinks it is the timelock leader or not. If I was initially not the leader, the metric I build will say in the name that I am not the leader - but if I become leader after that, even though it's the same method I need to generate one that says I am now the leader, and report that.

Reply:
I see, in that case the current cache won't quite work either because nothing triggers eviction.

Reply:
Right - yeah that's why I applied the do not merge label. The cache keys would need to have both the `Method` and `Map<String, String>` for the tags. We still pay an allocation to create this object, but I anticipate it will be quite a bit less than the `StringBuilder` issues we would have previously with the `MetricName` objects.